fs/io_zenfs.cc

```diff
@@ -659,6 +660,64 @@ size_t ZonedRandomAccessFile::GetUniqueId(char* id, size_t max_size) const {
   return zoneFile_->GetUniqueId(id, max_size);
 }
 
+ZenFSGCWorker::ZenFSGCWorker() { total_residue_ = 0; }
+
+void ZenFSGCWorker::CheckZoneValidResidualData() {
+  std::map<std::string, ZoneFile*>::iterator it;
+
+  fs->files_mtx_.lock();
+  for (it = fs->files_.begin(); it != fs->files_.end(); it++) {
+    ZoneFile* existFile;
+    existFile = it->second;
+
+    for (auto ext_it : existFile->extents_) {
+      Zone* zone_idx = ext_it->zone_;
+      // only care about the FULL zone.
+      if (!zone_idx->IsFull()) {
+        continue;
+      }
+
+      zone_residue[zone_idx] += ext_it->length_;
+      total_residue_ += ext_it->length_;
+      extent_list.push_back(ext_it);
+    }
+
+    files_moved_to_dst_zone.push_back(existFile);
+  }
+  fs->files_mtx_.unlock();
+}
+
+void ZenFSGCWorker::ZoneResetToReclaim() {
+  std::vector<Zone*>::iterator zone_it;
+  for (zone_it = merge_zone_list.begin(); zone_it != merge_zone_list.end();
+       zone_it++) {
+    Zone* zone_idx;
+    zone_idx = *zone_it;
+
+    IOStatus s;
+    s = zone_idx->Reset();
+    if (!s.ok()) {
+      // Debug(logger_, "Failed resetting zone when executing GC!");
+    }
+  }
+}
+
+std::vector<Zone*> ZenFSGCWorker::MarkZonesToMergeData() {
+  merge_zone_list = zbd_->GetReclaimZones();
+  return merge_zone_list;
+}
+
+std::vector<Zone*> ZenFSGCWorker::GetDestZoneToMoveValidData(
+    uint64_t ttl_residue) {
+  while (ttl_residue) {
+    Zone* zone = zbd_->AllocateZone(Env::WLTH_NOT_SET);
```

Comment:
Do we have any consideration on zone lifetime here? If leave this with not set, new extents with different lifetime will be pushed onto this zone, making it (possibly) harder to collect garbage.

Reply:
And what if zone allocation fails?